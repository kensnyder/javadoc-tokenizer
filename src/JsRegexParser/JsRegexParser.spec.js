const JsRegexParser = require('./JsRegexParser.js');
const pretty = require('pretty-var-export');
const fs = require('fs');
const fixtureJs1Function = fs.readFileSync(
	`${__dirname}/../../test-fixtures/js-1-function.js`,
	'utf8'
);
const fixtureJs2Functions = fs.readFileSync(
	`${__dirname}/../../test-fixtures/js-2-functions.js`,
	'utf8'
);

describe('JsRegexParser.getContext', () => {
	it('should find classes', () => {
		const parser = new JsRegexParser();
		const context = parser.getContext('class MyClass {');
		expect(context).toEqual({
			type: 'class',
			subtype: null,
			name: 'MyClass',
			isAsync: null,
			argsString: null,
			canAddDocgen: false,
		});
	});
	it('should find functions', () => {
		const parser = new JsRegexParser();
		const context = parser.getContext('function(a) {');
		expect(context).toEqual({
			type: 'function',
			subtype: null,
			name: '',
			isAsync: false,
			argsString: 'a',
			canAddDocgen: false,
		});
	});
	it('should find constructor', () => {
		const parser = new JsRegexParser();
		const context = parser.getContext('constructor(b = 2) {');
		expect(context).toEqual({
			type: 'function',
			subtype: 'constructor',
			name: 'constructor',
			isAsync: false,
			argsString: 'b = 2',
			canAddDocgen: false,
		});
	});
	it('should find exported functions', () => {
		const parser = new JsRegexParser();
		const context = parser.getContext('export function(c = "5") {');
		expect(context).toEqual({
			type: 'function',
			subtype: null,
			name: '',
			isAsync: false,
			argsString: 'c = "5"',
			canAddDocgen: false,
		});
	});
	it('should find default exported functions', () => {
		const parser = new JsRegexParser();
		const context = parser.getContext(
			'export default async function(d = {}) {'
		);
		expect(context).toEqual({
			type: 'function',
			subtype: null,
			name: '',
			isAsync: true,
			argsString: 'd = {}',
			canAddDocgen: false,
		});
	});
	it('should find assigned functions', () => {
		const parser = new JsRegexParser();
		const context = parser.getContext('const a = async function(e = []) {');
		expect(context).toEqual({
			type: 'function',
			subtype: 'variable',
			name: 'a',
			isAsync: true,
			argsString: 'e = []',
			canAddDocgen: true,
		});
	});
	it('should find property functions', () => {
		const parser = new JsRegexParser();
		const context = parser.getContext('a.b = function(f = null) {');
		expect(context).toEqual({
			type: 'function',
			subtype: 'variable',
			name: 'a.b',
			isAsync: false,
			argsString: 'f = null',
			canAddDocgen: true,
		});
	});
	it('should find class methods', () => {
		const parser = new JsRegexParser();
		const context = parser.getContext('async method(g = false) {');
		expect(context).toEqual({
			type: 'function',
			subtype: 'method',
			name: 'method',
			isAsync: true,
			argsString: 'g = false',
			canAddDocgen: false,
		});
	});
	it('should find class property methods', () => {
		const parser = new JsRegexParser();
		const context = parser.getContext('method: async (h = undefined) => {');
		expect(context).toEqual({
			type: 'function',
			subtype: 'method',
			name: 'method',
			isAsync: true,
			argsString: 'h = undefined',
			canAddDocgen: false,
		});
	});
	it('should find class properties', () => {
		const parser = new JsRegexParser();
		const context = parser.getContext('this.max = 5;');
		expect(context).toEqual({
			type: 'variable',
			subtype: 'property',
			name: 'this.max',
			isAsync: null,
			argsString: null,
			canAddDocgen: false,
		});
	});
});

describe('JsRegexParser.extractBlocks', () => {
	it('should find 1 comment block', () => {
		const parser = new JsRegexParser();
		const blocks = parser.extractBlocks(fixtureJs1Function);
		expect(blocks).toMatchSnapshot();
	});
	it('should find 2 comment blocks', () => {
		const parser = new JsRegexParser();
		const blocks = parser.extractBlocks(fixtureJs2Functions);
		expect(blocks).toMatchSnapshot();
	});
});
