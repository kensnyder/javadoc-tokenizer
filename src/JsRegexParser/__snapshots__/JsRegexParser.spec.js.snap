// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsRegexParser.extractBlocks should find 1 comment block 1`] = `
Array [
  Object {
    "argsString": "bytes, precision = 'auto'",
    "canAddDocgen": true,
    "code": "export default function bytesToText(bytes, precision = 'auto')",
    "comment": "/**
 * Convert numeric bytes to a rounded number with label
 * @example
 * bytesToText(23 * 1024 + 35); // 23.4 KB
 * @param {Number} bytes  The number of bytes
 * @param {Number|String} precision  The decimal precision or \\"auto\\"
 * @returns {String}
 */",
    "isAsync": false,
    "name": "bytesToText",
    "subtype": null,
    "type": "function",
  },
]
`;

exports[`JsRegexParser.extractBlocks should find 2 comment blocks 1`] = `
Array [
  Object {
    "argsString": "min, max",
    "canAddDocgen": true,
    "code": "export default function random(min, max)",
    "comment": "/**
 * Return a random integer between min and max inclusive
 * @param {Number} min  The lowest possible integer
 * @param {Number} max  The highest possible integer
 * @returns {Number}
 */",
    "isAsync": false,
    "name": "random",
    "subtype": null,
    "type": "function",
  },
  Object {
    "argsString": "min, max",
    "canAddDocgen": true,
    "code": "random.float = function (min, max)",
    "comment": "/**
 * Return a random float between min and max exclusive
 * @param {Number} min  The lowest possible value
 * @param {Number} max  The highest possible value
 * @returns {Number}
 */",
    "isAsync": false,
    "name": "random.float",
    "subtype": "variable",
    "type": "function",
  },
]
`;
